### PR TITLE
feat(pubsub): the streaming pull state machine

### PR DESCRIPTION
**Pull request by coryan**
Implements the state machine for streaming pulls. It initiates the
streaming pull request: creates the wrapper, then calls `Start()` and
the initial `Write()` both asynchronously. Then tracks if there is a
`Read()` and/or `Write()` call outstanding, queuing data for the next
`Write()` request if necessary. On an error, it waits for any
outstanding requests, then closes the stream gracefully, and resumes a
new streaming pull.

Part of the work for #5026 

Apologies for the monster CL, but I am not sure I can split this in any
sensible way. For what it is worth, most of the code is tests :blush: 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5190)

### CODE REVIEW

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.cc [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+  }
+  return os;
+}
+
+void StreamingSubscriptionBatchSource::ChangeState(
+    std::unique_lock<std::mutex> const&, StreamState s, char const* where,
+    char const* reason) {
+  GCP_LOG(DEBUG) << where << " (" << reason << ") " << stream_state_ << ":" << s
+                 << " read=" << pending_read_ << " write=" << pending_write_
+                 << " shutdown=" << shutdown_ << " stream=" << stream_.get()
```
Comment (devbww) [inline google/cloud/pubsub/internal/streaming_subscription_batch_source.cc:303]: Q. Is rendering an address meaningful?

**Review by devbww — APPROVED** [review]
> Apologies for the monster CL

I've done my best.

File: google/cloud/pubsub/integration_tests/subscriber_integration_test.cc [hunk 2]
```diff
@@ -139,6 +140,86 @@ TEST_F(SubscriberIntegrationTest, RawStub) {
...
+  std::set<std::string> expected_ids = [&] {
+    std::set<std::string> ids;
+    std::vector<future<StatusOr<std::string>>> message_ids;
+    for (int i = 0; i != kPublishCount; ++i) {
+      message_ids.push_back(publisher.Publish(
+          MessageBuilder{}.SetData("message-" + std::to_string(i)).Build()));
+    }
+    for (auto& id : message_ids) {
+      auto r = id.get();
+      EXPECT_STATUS_OK(r);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/integration_tests/subscriber_integration_test.cc:203]: Can't `ASSERT_STATUS_OK()` has an embedded `return;` in it.  But I should check `r` before using it.

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.cc]: No. I just wanted to know if it was `nullptr`, changed the debug logging to do that. PTAL.

**Comment by coryan** [comment]
@devbww if you want further changes just let me know.  I have merged this to keep the momentum.

**Pull request merged**